File: backend/ts_template/devdonalds.ts
import express, { Request, Response } from "express";

// ==== Type Definitions, feel free to add or modify ==========================
interface cookbookEntry {
  name: string;
  type: string;
}

interface requiredItem {
  name: string;
  quantity: number;
}

interface recipe extends cookbookEntry {
  requiredItems: requiredItem[];
}

interface ingredient extends cookbookEntry {
  cookTime: number;
}

// =============================================================================
// ==== HTTP Endpoint Stubs ====================================================
// =============================================================================
const app = express();
app.use(express.json());

// Store your recipes here!
const cookbook: Map<string, cookbookEntry> = new Map();

// Task 1 helper (don't touch)
app.post("/parse", (req:Request, res:Response) => {
  const { input } = req.body;

  const parsed_string = parse_handwriting(input)
  if (parsed_string == null) {
    res.status(400).send("this string is cooked");
    return;
  } 
  res.json({ msg: parsed_string });
  return;
  
});

// [TASK 1] ====================================================================
// Takes in a recipeName and returns it in a form that 
const parse_handwriting = (recipeName: string): string | null => {
  recipeName = recipeName
    .replace(/[-_]/g, " ")
    .replace(/[^a-zA-Z ]/g, "")
    .replace(/\s+/g, " ")
    .trim()
    .split(" ")
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
    .join(" ");

  return recipeName.length > 0 ? recipeName : null;
}

// [TASK 2] ====================================================================
// Endpoint that adds a CookbookEntry to your magical cookbook
app.post("/entry", (req:Request, res:Response) => {
  const { type, name, requiredItems, cookTime } = req.body;

  if (type !== "recipe" && type !== "ingredient") {
    return res.status(400).send({ error: "Type can only be recipe or ingredient." });
  }

  if (type === "ingredient" && cookTime < 0) {
    return res.status(400).send({ error: "cookTime can only be greater than or equal to 0." })
  }

  if (cookbook.has(name)) {
    return res.status(400).send({ error: "Entry names must be unique." });
  }

  if (type === "recipe") {
    const requiredSet = new Set();
    for (const r of requiredItems) {
      requiredSet.add(r.name);
    }

    if (requiredSet.size !== requiredItems.length) {
      return res.status(400).send({ error: "Recipe requiredItems can only have one element per name." });
    }
  }

  cookbook.set(name, req.body);
  return res.status(200).send({});
});

// [TASK 3] ====================================================================
// Endpoint that returns a summary of a recipe that corresponds to a query name
app.get("/summary", (req:Request, res:Response) => {
  const { name } = req.query;

  if (!cookbook.has(name)) {
    return res.status(400).send({ error: "A recipe with the corresponding name cannot be found." });
  }

  const entry = cookbook.get(name);

  if (entry.type !== "recipe") {
    return res.status(400).send({ error: "The searched name is NOT a recipe name (ie. an ingredient)." });
  }

  const dfs = (entry: recipe, multiplier: number = 1) => {
    let totalCookTime = 0;
    const totalIngredients = new Map();

    for (const requiredItem of entry.requiredItems) {
      const { name: itemName, quantity } = requiredItem;

      if (!cookbook.has(itemName)) {
        throw new Error("The recipe contains recipes or ingredients that aren't in the cookbook.");
      }

      const item = cookbook.get(itemName);
      const qty = quantity * multiplier;

      if (item.type === "ingredient") {
        totalCookTime += (item as ingredient).cookTime * qty;

        if (totalIngredients.has(itemName)) {
          totalIngredients.get(itemName).quantity += qty;
        } else {
          totalIngredients.set(itemName, { name: itemName, quantity: qty });
        }
      } else {
        const { cookTime, ingredients } = dfs(item as recipe, qty);

        totalCookTime += cookTime;

        for (const ingredient of ingredients.values()) {
          if (totalIngredients.has(ingredient.name)) {
            totalIngredients.get(ingredient.name).quantity += ingredient.quantity;
          } else {
            totalIngredients.set(ingredient.name, { ...ingredient });
          }
        }
      }
    }

    return { cookTime: totalCookTime, ingredients: totalIngredients };
  }

  try {
    const { cookTime, ingredients } = dfs(entry as recipe);
    return res.status(200).send({
      name,
      cookTime,
      ingredients: Array.from(ingredients.values())
    });
  } catch (err) {
    return res.status(400).send({ error: err.message });
  }
});

app.post("/reset", (req, res) => {
  cookbook.clear();
  res.status(200).send({});
});

// =============================================================================
// ==== DO NOT TOUCH ===========================================================
// =============================================================================
const port = 8080;
app.listen(port, () => {
  console.log(`Running on: http://127.0.0.1:8080`);
});

File: backend/py_template/devdonalds.py
from dataclasses import dataclass
from typing import List, Dict, Union
from flask import Flask, request, jsonify
import re

# ==== Type Definitions, feel free to add or modify ===========================
@dataclass
class CookbookEntry:
	name: str

@dataclass
class RequiredItem():
	name: str
	quantity: int

@dataclass
class Recipe(CookbookEntry):
	required_items: List[RequiredItem]

@dataclass
class Ingredient(CookbookEntry):
	cook_time: int


# =============================================================================
# ==== HTTP Endpoint Stubs ====================================================
# =============================================================================
app = Flask(__name__)

# Store your recipes here!
cookbook = {}

# Task 1 helper (don't touch)
@app.route("/parse", methods=['POST'])
def parse():
	data = request.get_json()
	recipe_name = data.get('input', '')
	parsed_name = parse_handwriting(recipe_name)
	if parsed_name is None:
		return 'Invalid recipe name', 400
	return jsonify({'msg': parsed_name}), 200

# [TASK 1] ====================================================================
# Takes in a recipeName and returns it in a form that 
def parse_handwriting(recipeName: str) -> Union[str | None]:
	recipeName = recipeName.replace("-", " ").replace("_", " ")
	recipeName = re.sub(r'[^a-zA-Z ]+', "", recipeName)
	recipeName = re.sub(r'\s+', " ", recipeName)
	recipeName = recipeName.strip()

	return recipeName.title() if len(recipeName) > 0 else None


# [TASK 2] ====================================================================
# Endpoint that adds a CookbookEntry to your magical cookbook
@app.route('/entry', methods=['POST'])
def create_entry():
	data = request.json
	entryType, name, requiredItems, cookTime = (
		data.get("type"),
		data.get("name"),
		data.get("requiredItems"),
		data.get("cookTime"),
	)

	if entryType not in ["recipe", "ingredient"]:
		return "Type can only be recipe or ingredient.", 400

	if entryType == "ingredient" and cookTime < 0:
		return "cookTime can only be greater than or equal to 0.", 400
	
	if name in cookbook:
		return "Entry names must be unique.", 400

	if entryType == "recipe":
		requiredSet = set()
		for item in requiredItems:
			requiredSet.add(item.get("name"))

		if len(requiredSet) != len(requiredItems):
			return "Recipe requiredItems can only have one element per name.", 400

	cookbook[name] = data
	return {}, 200


# [TASK 3] ====================================================================
# Endpoint that returns a summary of a recipe that corresponds to a query name
@app.route('/summary', methods=['GET'])
def summary():
	name = request.args.get("name")

	if name not in cookbook:
		return "A recipe with the corresponding name cannot be found.", 400

	entry = cookbook[name]

	if entry.get("type") != "recipe":
		return "The searched name is NOT a recipe name (ie. an ingredient).", 400

	def dfs(entry, multiplier=1):
		totalCookTime = 0
		totalIngredients = {}

		for requiredItem in entry.get("requiredItems"):
			itemName = requiredItem.get("name")
			quantity = requiredItem.get("quantity")

			if itemName not in cookbook:
				raise ValueError("The recipe contains recipes or ingredients that aren't in the cookbook.")

			item = cookbook[itemName]
			qty = quantity * multiplier

			if item.get("type") == "ingredient":
				totalCookTime += item.get("cookTime") * qty

				if itemName in totalIngredients:
					totalIngredients[itemName]["quantity"] += qty
				else:
					totalIngredients[itemName] = {"name": itemName, "quantity": qty}
			else:
				cookTime, ingredients = dfs(item, qty)

				totalCookTime += cookTime

				for ingredientName, ingredientData in ingredients.items():
					if ingredientName in totalIngredients:
							totalIngredients[ingredientName]["quantity"] += ingredientData["quantity"]
					else:
							totalIngredients[ingredientName] = ingredientData.copy()

		return totalCookTime, totalIngredients

	try:
		cookTime, ingredients = dfs(entry)
		return jsonify({
			"name": name,
			"cookTime": cookTime,
			"ingredients": list(ingredients.values())
		}), 200
	except ValueError as e:
		return str(e), 400

@app.route('/reset', methods=['POST'])
def reset():
	cookbook.clear()
	return {}, 200

# =============================================================================
# ==== DO NOT TOUCH ===========================================================
# =============================================================================

if __name__ == '__main__':
	app.run(debug=True, port=8080)
